File: clients_sdk/typescript/src/pipelines/talk.ts
/**
 * BudTalk Pipeline
 * Bidirectional voice pipeline combining STT + TTS
 */

import type { STTConfig, TTSConfig, LiveKitConfig } from '../types/config.js';
import type { FeatureFlags } from '../types/features.js';
import type { MetricsSummary } from '../types/metrics.js';
import type { TranscriptEvent, AudioEvent } from '../ws/events.js';
import { BasePipeline, type BasePipelineConfig } from './base.js';
import { AudioRecorder, type RecorderConfig } from '../audio/recorder.js';
import { PCMPlayer, type PlayerConfig } from '../audio/player.js';
import { VAD, type VADConfig } from '../audio/vad.js';
import { getMetricsCollector } from '../metrics/collector.js';

/**
 * BudTalk configuration
 */
export interface BudTalkConfig extends BasePipelineConfig {
  /** STT configuration */
  stt?: Partial<STTConfig>;
  /** TTS configuration */
  tts?: Partial<TTSConfig>;
  /** LiveKit configuration for room-based communication */
  livekit?: LiveKitConfig;
  /** VAD configuration */
  vadConfig?: VADConfig;
  /** Recorder configuration */
  recorderConfig?: RecorderConfig;
  /** Player configuration */
  playerConfig?: PlayerConfig;
  /** Whether to auto-play received audio (default: true) */
  autoPlay?: boolean;
  /** Whether to auto-record from microphone (default: false) */
  autoRecord?: boolean;
  /** Interrupt TTS when user starts speaking (default: true) */
  bargeIn?: boolean;
}

/**
 * Conversation turn
 */
export interface ConversationTurn {
  /** Who spoke: 'user' or 'assistant' */
  role: 'user' | 'assistant';
  /** The text content */
  text: string;
  /** Timestamp */
  timestamp: number;
  /** Whether it's complete */
  isFinal: boolean;
  /** Audio duration if available */
  audioDurationMs?: number;
}

/**
 * BudTalk - Bidirectional Voice Pipeline
 */
export class BudTalk extends BasePipeline {
  private sttConfig: STTConfig;
  private ttsConfig: TTSConfig;
  private livekitConfig?: LiveKitConfig;
  private recorder: AudioRecorder | null = null;
  private player: PCMPlayer | null = null;
  private vad: VAD | null = null;
  private autoPlay: boolean;
  private autoRecord: boolean;
  private bargeIn: boolean;
  private isListening = false;
  private isSpeaking = false;
  private conversationHistory: ConversationTurn[] = [];
  private currentUserUtterance = '';
  private e2eLatencyStart: number | null = null;

  constructor(config: BudTalkConfig) {
    const sttConfig: STTConfig = {
      provider: config.stt?.provider ?? 'deepgram',
      language: config.stt?.language ?? 'en-US',
      model: config.stt?.model,
      sampleRate: config.stt?.sampleRate ?? 16000,
      encoding: config.stt?.encoding ?? 'linear16',
      channels: config.stt?.channels ?? 1,
      interimResults: config.stt?.interimResults ?? true,
      punctuate: config.stt?.punctuate ?? true,
      profanityFilter: config.stt?.profanityFilter ?? false,
      smartFormat: config.stt?.smartFormat ?? true,
      diarize: config.stt?.diarize ?? false,
    };

    const ttsConfig: TTSConfig = {
      provider: config.tts?.provider ?? 'deepgram',
      voice: config.tts?.voice,
      voiceId: config.tts?.voiceId,
      model: config.tts?.model,
      sampleRate: config.tts?.sampleRate ?? 24000,
      audioFormat: config.tts?.audioFormat ?? 'linear16',
      speed: config.tts?.speed,
      pitch: config.tts?.pitch,
      volume: config.tts?.volume,
    };

    super({
      ...config,
      sessionConfig: {
        stt: sttConfig,
        tts: ttsConfig,
        livekit: config.livekit,
      },
    });

    this.sttConfig = sttConfig;
    this.ttsConfig = ttsConfig;
    this.livekitConfig = config.livekit;
    this.autoPlay = config.autoPlay ?? true;
    this.autoRecord = config.autoRecord ?? false;
    this.bargeIn = config.bargeIn ?? true;

    // Setup VAD
    if (config.features?.vad !== false) {
      this.vad = new VAD({
        sampleRate: sttConfig.sampleRate,
        ...config.vadConfig,
      });

      this.vad.setHandlers({
        onSpeechStart: (buffer) => this.handleSpeechStart(buffer),
        onSpeechEnd: (duration) => this.handleSpeechEnd(duration),
      });
    }

    // Setup player
    if (this.autoPlay) {
      this.player = new PCMPlayer({
        sampleRate: ttsConfig.sampleRate,
        ...config.playerConfig,
      });
    }

    this.setupEventHandlers();
  }

  /**
   * Setup internal event handlers
   */
  private setupEventHandlers(): void {
    // Handle transcripts
    this.session.on('transcript', (e) => {
      this.handleTranscript(e);
    });

    // Handle audio
    this.session.on('audio', (e) => {
      this.handleAudio(e);
    });

    // Handle speaking state
    this.session.on('speaking', (e) => {
      this.isSpeaking = e.speaking;

      if (e.speaking && this.e2eLatencyStart) {
        // Record E2E latency (from user speech end to bot speech start)
        const latency = Date.now() - this.e2eLatencyStart;
        getMetricsCollector().record('e2e.latency', latency);
        this.e2eLatencyStart = null;
      }

      this.emitter.emit('speaking', e);
    });

    // Handle listening state
    this.session.on('listening', (e) => {
      this.emitter.emit('listening', e);
    });
  }

  /**
   * Handle VAD speech start
   */
  private handleSpeechStart(buffer: Float32Array): void {
    this.isListening = true;
    this.emitter.emit('listening', { listening: true, timestamp: Date.now() });

    // Barge-in: interrupt current TTS
    if (this.bargeIn && this.isSpeaking) {
      this.interrupt();
    }

    // Send buffered pre-speech audio
    if (buffer.length > 0) {
      const int16 = new Int16Array(buffer.length);
      for (let i = 0; i < buffer.length; i++) {
        const sample = Math.max(-1, Math.min(1, buffer[i]!));
        int16[i] = sample < 0 ? sample * 32768 : sample * 32767;
      }
      this.session.sendAudio(int16.buffer);
    }
  }

  /**
   * Handle VAD speech end
   */
  private handleSpeechEnd(_duration: number): void {
    this.isListening = false;
    this.emitter.emit('listening', { listening: false, timestamp: Date.now() });

    // Mark E2E latency start
    this.e2eLatencyStart = Date.now();

    // Flush to finalize transcription
    this.session.flush();
  }

  /**
   * Handle transcript events
   */
  private handleTranscript(event: TranscriptEvent): void {
    if (event.isFinal) {
      // Add to conversation history
      this.conversationHistory.push({
        role: 'user',
        text: event.text,
        timestamp: Date.now(),
        isFinal: true,
      });
      this.currentUserUtterance = '';
    } else {
      this.currentUserUtterance = event.text;
    }

    this.emitter.emit('transcript', event);
  }

  /**
   * Handle audio events
   */
  private handleAudio(event: AudioEvent): void {
    this.emitter.emit('audio', event);

    if (this.autoPlay && this.player) {
      this.player.addPCM(event.audio);
    }
  }

  /**
   * Connect and optionally start recording
   */
  async connect(): Promise<void> {
    await super.connect();

    // Initialize player
    if (this.autoPlay && this.player) {
      await this.player.initialize();
    }

    // Auto-record if enabled
    if (this.autoRecord) {
      await this.startListening();
    }
  }

  /**
   * Send audio data
   */
  sendAudio(audio: ArrayBuffer | Int16Array | Uint8Array): void {
    if (!this.isConnected()) {
      throw new Error('Not connected. Call connect() first.');
    }

    const buffer = audio instanceof Int16Array
      ? audio.buffer as ArrayBuffer
      : audio instanceof Uint8Array
        ? audio.buffer as ArrayBuffer
        : audio;

    // Process through VAD if enabled
    if (this.vad) {
      const int16 = new Int16Array(buffer);
      const float32 = new Float32Array(int16.length);
      for (let i = 0; i < int16.length; i++) {
        float32[i] = int16[i]! / 32768;
      }

      const events = this.vad.process(float32);

      for (const event of events) {
        if (event.state === 'speech') {
          this.session.sendAudio(buffer);
          break;
        }
      }
    } else {
      this.session.sendAudio(buffer);
    }
  }

  /**
   * Speak text
   */
  async speak(text: string, options?: {
    voice?: string;
    voiceId?: string;
    provider?: string;
    model?: string;
    speed?: number;
    pitch?: number;
  }): Promise<void> {
    // Add to conversation history
    this.conversationHistory.push({
      role: 'assistant',
      text,
      timestamp: Date.now(),
      isFinal: true,
    });

    this.session.speak(text, options);
  }

  /**
   * Start listening from microphone
   */
  async startListening(recorderConfig?: RecorderConfig): Promise<void> {
    if (this.isListening && this.recorder) return;

    if (!this.isConnected()) {
      await this.connect();
    }

    this.recorder = new AudioRecorder({
      sampleRate: this.sttConfig.sampleRate,
      ...recorderConfig,
    });

    this.recorder.setHandlers({
      onData: (data) => this.sendAudio(data),
      onError: (error) => {
        this.emitter.emit('error', {
          code: 'RECORDER_ERROR',
          message: error.message,
          recoverable: true,
          raw: { type: 'error', code: 'RECORDER_ERROR', message: error.message },
        });
      },
    });

    await this.recorder.start();
    this.isListening = true;
  }

  /**
   * Stop listening
   */
  stopListening(): void {
    if (!this.recorder) return;

    this.recorder.stop();
    this.isListening = false;
    this.session.flush();
  }

  /**
   * Interrupt current TTS
   */
  interrupt(): void {
    super.interrupt();
    this.player?.stop();
  }

  /**
   * Get conversation history
   */
  getConversationHistory(): ConversationTurn[] {
    return [...this.conversationHistory];
  }

  /**
   * Clear conversation history
   */
  clearConversationHistory(): void {
    this.conversationHistory = [];
  }

  /**
   * Check if listening
   */
  getIsListening(): boolean {
    return this.isListening;
  }

  /**
   * Check if speaking
   */
  getIsSpeaking(): boolean {
    return this.isSpeaking;
  }

  /**
   * Update STT config
   */
  updateSTTConfig(config: Partial<STTConfig>): void {
    this.sttConfig = { ...this.sttConfig, ...config };
    this.session.updateSTTConfig(config);
  }

  /**
   * Update TTS config
   */
  updateTTSConfig(config: Partial<TTSConfig>): void {
    this.ttsConfig = { ...this.ttsConfig, ...config };
    this.session.updateTTSConfig(config);
  }

  /**
   * Set barge-in enabled
   */
  setBargeIn(enabled: boolean): void {
    this.bargeIn = enabled;
  }

  /**
   * Set playback volume
   */
  setVolume(gainDb: number): void {
    this.player?.setGain(gainDb);
  }

  /**
   * Dispose resources
   */
  async dispose(): Promise<void> {
    this.stopListening();

    if (this.recorder) {
      await this.recorder.dispose();
      this.recorder = null;
    }

    if (this.player) {
      await this.player.dispose();
      this.player = null;
    }

    await this.disconnect();
  }
}

/**
 * Create a BudTalk instance
 */
export function createBudTalk(config: BudTalkConfig): BudTalk {
  return new BudTalk(config);
}

File: clients_sdk/python/bud_foundry/pipelines/talk.py
"""
BudTalk - Bidirectional voice pipeline
"""

from typing import Any, AsyncIterator, Callable, Optional, Union
from dataclasses import dataclass

from ..types import STTConfig, TTSConfig, STTResult, AudioEvent, FeatureFlags
from ..ws.session import WebSocketSession, SessionMetrics, ReconnectConfig


@dataclass
class TalkEvent:
    """Event from a Talk session."""

    type: str
    """Event type: transcript, audio, message, error, playback_complete"""

    transcript: Optional[STTResult] = None
    """Transcript result (if type is 'transcript')"""

    audio: Optional[AudioEvent] = None
    """Audio event (if type is 'audio')"""

    message: Optional[dict[str, Any]] = None
    """Message data (if type is 'message')"""

    error: Optional[Exception] = None
    """Error (if type is 'error')"""


class BudTalk:
    """Bidirectional voice pipeline combining STT and TTS."""

    def __init__(
        self,
        url: str,
        api_key: Optional[str] = None,
    ):
        """
        Initialize Talk pipeline.

        Args:
            url: WebSocket URL of the gateway
            api_key: Optional API key for authentication
        """
        self.url = url
        self.api_key = api_key

    def create(
        self,
        stt: Optional[Union[STTConfig, dict[str, Any]]] = None,
        tts: Optional[Union[TTSConfig, dict[str, Any]]] = None,
        livekit: Optional[dict[str, Any]] = None,
        features: Optional[FeatureFlags] = None,
        reconnect: Optional[ReconnectConfig] = None,
    ) -> "TalkSession":
        """
        Create a Talk session.

        Args:
            stt: STT configuration or dict
            tts: TTS configuration or dict
            livekit: LiveKit configuration
            features: Feature flags
            reconnect: Reconnection configuration

        Returns:
            Talk session
        """
        # Convert dicts to config objects
        stt_config: Optional[STTConfig] = None
        if stt is not None:
            if isinstance(stt, dict):
                stt_config = STTConfig(
                    provider=stt.get("provider", "deepgram"),
                    language=stt.get("language", "en-US"),
                    model=stt.get("model"),
                    sample_rate=stt.get("sample_rate", 16000),
                    channels=stt.get("channels", 1),
                    encoding=stt.get("encoding", "linear16"),
                    interim_results=stt.get("interim_results", True),
                    punctuate=stt.get("punctuate", True),
                )
            else:
                stt_config = stt

        # Apply feature flags to STT config
        if stt_config and features:
            stt_config.interim_results = features.interim_results
            stt_config.punctuate = features.punctuation
            stt_config.profanity_filter = features.profanity_filter
            stt_config.smart_format = features.smart_format
            stt_config.diarize = features.speaker_diarization

        tts_config: Optional[TTSConfig] = None
        if tts is not None:
            if isinstance(tts, dict):
                tts_config = TTSConfig(
                    provider=tts.get("provider", "deepgram"),
                    voice=tts.get("voice"),
                    voice_id=tts.get("voice_id"),
                    model=tts.get("model"),
                    sample_rate=tts.get("sample_rate", 24000),
                )
            else:
                tts_config = tts

        return TalkSession(
            url=self.url,
            api_key=self.api_key,
            stt_config=stt_config,
            tts_config=tts_config,
            livekit_config=livekit,
            features=features,
            reconnect=reconnect,
        )

    async def connect(
        self,
        stt: Optional[Union[STTConfig, dict[str, Any]]] = None,
        tts: Optional[Union[TTSConfig, dict[str, Any]]] = None,
        livekit: Optional[dict[str, Any]] = None,
        features: Optional[FeatureFlags] = None,
        reconnect: Optional[ReconnectConfig] = None,
    ) -> "TalkSession":
        """
        Create and connect a Talk session.

        Args:
            stt: STT configuration
            tts: TTS configuration
            livekit: LiveKit configuration
            features: Feature flags
            reconnect: Reconnection configuration

        Returns:
            Connected Talk session
        """
        session = self.create(
            stt=stt,
            tts=tts,
            livekit=livekit,
            features=features,
            reconnect=reconnect,
        )
        await session.connect()
        return session


class TalkSession:
    """Active Talk session for bidirectional voice communication."""

    def __init__(
        self,
        url: str,
        api_key: Optional[str] = None,
        stt_config: Optional[STTConfig] = None,
        tts_config: Optional[TTSConfig] = None,
        livekit_config: Optional[dict[str, Any]] = None,
        features: Optional[FeatureFlags] = None,
        reconnect: Optional[ReconnectConfig] = None,
    ):
        """
        Initialize Talk session.

        Args:
            url: WebSocket URL
            api_key: API key
            stt_config: STT configuration
            tts_config: TTS configuration
            livekit_config: LiveKit configuration
            features: Feature flags
            reconnect: Reconnection config
        """
        self.stt_config = stt_config
        self.tts_config = tts_config
        self.livekit_config = livekit_config
        self.features = features

        self._session = WebSocketSession(
            url=url,
            api_key=api_key,
            stt_config=stt_config,
            tts_config=tts_config,
            livekit_config=livekit_config,
            reconnect=reconnect,
        )

        self._event_handlers: dict[str, list[Callable[..., Any]]] = {}

        # Wire up internal handlers
        self._session.on("transcript", self._on_transcript)
        self._session.on("audio", self._on_audio)
        self._session.on("message", self._on_message)
        self._session.on("error", self._on_error)
        self._session.on("playback_complete", self._on_playback_complete)

    def _on_transcript(self, result: STTResult) -> None:
        """Handle transcript events."""
        event = TalkEvent(type="transcript", transcript=result)
        self._emit("transcript", result)
        self._emit("event", event)

    def _on_audio(self, audio: AudioEvent) -> None:
        """Handle audio events."""
        event = TalkEvent(type="audio", audio=audio)
        self._emit("audio", audio)
        self._emit("event", event)

    def _on_message(self, message: dict[str, Any]) -> None:
        """Handle message events."""
        event = TalkEvent(type="message", message=message)
        self._emit("message", message)
        self._emit("event", event)

    def _on_error(self, error: Exception) -> None:
        """Handle error events."""
        event = TalkEvent(type="error", error=error)
        self._emit("error", error)
        self._emit("event", event)

    def _on_playback_complete(self, timestamp: Any) -> None:
        """Handle playback complete events."""
        event = TalkEvent(type="playback_complete")
        self._emit("playback_complete")
        self._emit("event", event)

    def _emit(self, event: str, *args: Any) -> None:
        """Emit an event."""
        handlers = self._event_handlers.get(event, [])
        for handler in handlers:
            try:
                handler(*args)
            except Exception:
                pass

    @property
    def connected(self) -> bool:
        """Whether the session is connected."""
        return self._session.connected

    @property
    def stream_id(self) -> Optional[str]:
        """Get the stream ID."""
        return self._session.stream_id

    def on(self, event: str, handler: Callable[..., Any]) -> None:
        """
        Register an event handler.

        Args:
            event: Event name (transcript, audio, message, error, event, etc.)
            handler: Event handler function
        """
        if event not in self._event_handlers:
            self._event_handlers[event] = []
        self._event_handlers[event].append(handler)

    def off(self, event: str, handler: Optional[Callable[..., Any]] = None) -> None:
        """Remove an event handler."""
        if event in self._event_handlers:
            if handler is None:
                self._event_handlers[event].clear()
            elif handler in self._event_handlers[event]:
                self._event_handlers[event].remove(handler)

    async def connect(self, timeout: float = 10.0) -> None:
        """Connect to the gateway."""
        await self._session.connect(timeout=timeout)

    async def disconnect(self) -> None:
        """Disconnect from the gateway."""
        await self._session.disconnect()

    async def send_audio(self, audio: Union[bytes, bytearray]) -> None:
        """
        Send audio data for transcription.

        Args:
            audio: PCM audio data (16-bit signed integer, mono)
        """
        await self._session.send_audio(audio)

    async def speak(
        self,
        text: str,
        flush: bool = False,
        allow_interruption: bool = True,
    ) -> None:
        """
        Synthesize and stream text to speech.

        Args:
            text: Text to synthesize
            flush: Whether to flush the TTS buffer immediately
            allow_interruption: Whether this TTS can be interrupted
        """
        await self._session.speak(text, flush=flush, allow_interruption=allow_interruption)

    async def clear(self) -> None:
        """Clear/stop current TTS playback."""
        await self._session.clear()

    async def send_message(
        self,
        message: str,
        role: str = "user",
        topic: Optional[str] = None,
    ) -> None:
        """
        Send a data message to other participants.

        Args:
            message: Message content
            role: Message role (user, assistant, system)
            topic: Optional topic/channel
        """
        await self._session.send_message(message, role=role, topic=topic)

    async def sip_transfer(self, transfer_to: str) -> None:
        """
        Transfer a SIP call.

        Args:
            transfer_to: Phone number to transfer to
        """
        await self._session.sip_transfer(transfer_to)

    def get_metrics(self) -> SessionMetrics:
        """Get session metrics."""
        return self._session.get_metrics()

    def reset_metrics(self) -> None:
        """Reset session metrics."""
        self._session.reset_metrics()

    async def __aiter__(self) -> AsyncIterator[TalkEvent]:
        """Iterate over all events."""
        async for message in self._session:
            msg_type = message.get("type")

            if msg_type == "transcript":
                yield TalkEvent(type="transcript", transcript=message["result"])
            elif msg_type == "audio":
                yield TalkEvent(type="audio", audio=message["audio"])
            elif msg_type == "message":
                yield TalkEvent(type="message", message=message["data"])
            elif msg_type == "error":
                yield TalkEvent(type="error", error=message["error"])
            elif msg_type == "playback_complete":
                yield TalkEvent(type="playback_complete")

    async def __aenter__(self) -> "TalkSession":
        """Async context manager entry."""
        await self.connect()
        return self

    async def __aexit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        """Async context manager exit."""
        await self.disconnect()
